core/src/main/java/com/alibaba/alink/operator/batch/dataproc/StratifiedSampleWithSizeBatchOp.java

```diff
@@ -0,0 +1,209 @@
+package com.alibaba.alink.operator.batch.dataproc;
+
+import com.alibaba.alink.operator.batch.BatchOperator;
+import com.alibaba.alink.params.dataproc.SampleSeedParam;
+import com.alibaba.alink.params.dataproc.SampleGroupColumnParam;
+import com.alibaba.alink.params.dataproc.SampleSizeParams;
+import com.alibaba.alink.params.dataproc.SampleWithReplacementParams;
+import com.google.common.base.Joiner;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.flink.api.common.functions.GroupReduceFunction;
+import org.apache.flink.api.java.operators.GroupReduceOperator;
+import org.apache.flink.api.java.operators.UnsortedGrouping;
+import org.apache.flink.api.java.sampling.IntermediateSampleData;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.XORShiftRandom;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+/**
+ * @auth：baijingjing
+ * @createDatetime: 2020/7/7
+ * @desc：fencengcaiyang main class
+ */
+public final class StratifiedSampleWithSizeBatchOp extends BatchOperator<StratifiedSampleWithSizeBatchOp>
+        implements SampleGroupColumnParam<StratifiedSampleWithSizeBatchOp>,
+                   SampleSizeParams<StratifiedSampleWithSizeBatchOp>,
+                   SampleSeedParam<StratifiedSampleWithSizeBatchOp>,
+                   SampleWithReplacementParams<StratifiedSampleWithSizeBatchOp> {
+
+    public StratifiedSampleWithSizeBatchOp() {
+        this(new Params());
+    }
+
+    public StratifiedSampleWithSizeBatchOp(Params params){
+        super(params.set(SEED, new Random().nextLong()));
+    }
+
+    public StratifiedSampleWithSizeBatchOp(String groupKey, String size) {
+        this(groupKey, size, false);
+    }
+    public StratifiedSampleWithSizeBatchOp(String groupKey, String size, boolean withReplacement) {
+        this(new Params()
+             .set(GROUP_COL, groupKey)
+             .set(SIZE, size)
+             .set(WITH_REPLACEMENT, withReplacement));
+    }
+
+    @Override
+    public StratifiedSampleWithSizeBatchOp linkFrom(BatchOperator<?>... inputs) {
+        BatchOperator<?> in = checkAndGetFirst(inputs);
+        // compute index of group key
+        int index = computeGroupCloIndex(in.getColNames(), getGroupCol());
+
+        UnsortedGrouping<Row> groupingOperator = in.getDataSet().groupBy(index);
+
+        String size = getSize();
+        //step2: count total number of records per group
+        try {
+            List<Tuple2<Object, Long>> totalNumberPerGroupList = groupingOperator.reduceGroup(new CountPerGroupFunction(index)).collect();
```

Comment:
Do not use the `collect()` if it is not necessary because `collect()` will break the user process. 

Use the `withBroadcastSet()` in the `DataSet` to brocadcast the dataset to all worker and do the function logic on the `open()` function of `GroupReduceFunction `

Another saying is that I think the `totalNumberPerGroupList` in this case may be not necessary